File: frontend-next/src/types/mutualFund.ts
export interface MFHoldingRow {
  change_type: string;
  fund_name: string;
  stock_name: string;
  mf_schcode: number;
  co_code: number;
  perc_aum: number;
  perc_aum_prev: number;
  share_count: number;
  share_count_prev: number;
  mkt_value: number;
  mkt_value_prev: number;
}

export interface MFHoldingSummary {
  new_entries: number;
  modified: number;
  unchanged: number;
  removed: number;
  total_funds: number;
}

export interface MFHoldingsResponse {
  month: string;
  prev_month: string;
  summary: MFHoldingSummary;
  rows: MFHoldingRow[];
}

export interface MFBuySellTrendPoint {
  month: string;
  buy_value: number;
  sell_value: number;
}

export interface MFNetValueItem {
  name: string;
  net_value: number;
}

export interface MFBuySellResponse {
  total_buy: number;
  total_sell: number;
  trend: MFBuySellTrendPoint[];
  by_stock: MFNetValueItem[];
  by_sector: MFNetValueItem[];
}

export interface MFPopularStock {
  name: string;
  count: number;
}

export interface MFInsightsResponse {
  month: string;
  most_popular: MFPopularStock[];
  least_popular: MFPopularStock[];
}

export interface MFAssetAllocationItem {
  fund_name: string;
  equity: number;
  debt: number;
  cash: number;
  misc: number;
}

export interface MFAssetAllocationResponse {
  month: string;
  items: MFAssetAllocationItem[];
}

export interface MFFiltersResponse {
  available_months: string[];
  fund_names: string[];
  stock_names: string[];
  categories: string[];
}

File: backend/schemas/mutual_fund.py
"""Pydantic schemas for Mutual Fund Holdings endpoints."""

from pydantic import BaseModel


class MFHoldingRow(BaseModel):
    change_type: str  # "New Entry", "Modified", "Unchanged", "Removed"
    fund_name: str
    stock_name: str
    mf_schcode: int
    co_code: int
    perc_aum: float  # current month % AUM
    perc_aum_prev: float  # previous month % AUM
    share_count: int  # current month shares
    share_count_prev: int  # previous month shares
    mkt_value: float  # current month market value (Cr)
    mkt_value_prev: float  # previous month market value (Cr)


class MFHoldingSummary(BaseModel):
    new_entries: int
    modified: int
    unchanged: int
    removed: int
    total_funds: int


class MFHoldingsResponse(BaseModel):
    month: str
    prev_month: str
    summary: MFHoldingSummary
    rows: list[MFHoldingRow]


class MFBuySellTrendPoint(BaseModel):
    month: str
    buy_value: float
    sell_value: float


class MFNetValueItem(BaseModel):
    name: str
    net_value: float


class MFBuySellResponse(BaseModel):
    total_buy: float
    total_sell: float
    trend: list[MFBuySellTrendPoint]
    by_stock: list[MFNetValueItem]
    by_sector: list[MFNetValueItem]


class MFPopularStock(BaseModel):
    name: str
    count: int


class MFInsightsResponse(BaseModel):
    month: str
    most_popular: list[MFPopularStock]
    least_popular: list[MFPopularStock]


class MFAssetAllocationItem(BaseModel):
    fund_name: str
    equity: float
    debt: float
    cash: float
    misc: float


class MFAssetAllocationResponse(BaseModel):
    month: str
    items: list[MFAssetAllocationItem]


class MFFiltersResponse(BaseModel):
    available_months: list[str]
    fund_names: list[str]
    stock_names: list[str]
    categories: list[str]
